Packages/MIES/MIES_SweepFormula.ipf

```diff
@@ -0,0 +1,597 @@
+					break
+				case STATE_COLLECT:
+				case STATE_DEFAULT:
+					action = ACTION_COLLECT
+					break
+				default:
+					ASSERT(0, "Encountered undefined transition " + num2str(state))
+			endswitch
+			lastState = state
+		endif
+
+		// action
+		switch(action)
+			case ACTION_COLLECT:
+				token += formula[i]
+			case ACTION_SKIP:
+				continue
+			case ACTION_FUNCTION:
+				tempPath = jsonPath
+				if(JSON_GetType(jsonID, jsonPath) == JSON_ARRAY)
+					JSON_AddObjects(jsonID, jsonPath)
+					tempPath += "/" + num2str(JSON_GetArraySize(jsonID, jsonPath) - 1)
+				endif
+				tempPath += "/"
+				parenthesisStart = strsearch(token, "(", 0, 0)
+				tempPath += EscapeJsonPath(token[0, parenthesisStart - 1])
+				jsonIDdummy = FormulaParser(token[parenthesisStart + 1, inf])
+				if(JSON_GetType(jsonIDdummy, "") != JSON_ARRAY)
+					JSON_AddTreeArray(jsonID, tempPath)
+				endif
+				JSON_AddJSON(jsonID, tempPath, jsonIDdummy)
+				JSON_Release(jsonIDdummy)
+				break
+			case ACTION_PARENTHESIS:
+				JSON_AddJSON(jsonID, jsonPath, FormulaParser(token[1, inf]))
+				break
+			case ACTION_HIGHERORDER:
+				lastCalculation = state
+				if(!!cmpstr(token, ""))
+					JSON_AddJSON(jsonID, jsonPath, FormulaParser(token))
+				endif
+				jsonPath = EscapeJsonPath(formula[i])
+				if(!cmpstr(jsonPath, ",") || !cmpstr(jsonPath, "]"))
+					jsonPath = ""
+				endif
+				jsonIDdummy = jsonID
+				jsonID = JSON_New()
+				JSON_AddTreeArray(jsonID, jsonPath)
+				JSON_AddJSON(jsonID, jsonPath, jsonIDdummy)
+				JSON_Release(jsonIDdummy)
+				break
+			case ACTION_ARRAY:
+				ASSERT(!cmpstr(token[0], "["), "Encountered array ending without array start.")
+				jsonIDdummy = FormulaParser(token[1,inf])
+				if(JSON_GetType(jsonIDdummy, "") != JSON_ARRAY)
+					JSON_AddTreeArray(jsonID, jsonPath)
+				endif
+				JSON_AddJSON(jsonID, jsonPath, jsonIDdummy)
+				JSON_Release(jsonIDdummy)
+				break
+			case ACTION_CALCULATION:
+				if(JSON_GetType(jsonID, jsonPath) == JSON_ARRAY)
+					JSON_AddObjects(jsonID, jsonPath) // prepare for decent
+					jsonPath += "/" + num2str(JSON_GetArraySize(jsonID, jsonPath) - 1)
+				endif
+				jsonPath += "/" + EscapeJsonPath(formula[i])
+			case ACTION_ARRAYELEMENT:
+				JSON_AddTreeArray(jsonID, jsonPath)
+				lastCalculation = state
+			case ACTION_SAMECALCULATION:
+			default:
+				if(strlen(token) > 0)
+					JSON_AddJSON(jsonID, jsonPath, FormulaParser(token))
+				endif
+		endswitch
+		token = ""
+	endfor
+
+	// last element (recursion)
+	if(!cmpstr(token, formula))
+		if(GrepString(token, "^(?i)[0-9]+(?:\.[0-9]+)?(?:[\+-]?E[0-9]+)?$"))
+			JSON_AddVariable(jsonID, jsonPath, str2num(formula))
+		else
+			JSON_AddString(jsonID, jsonPath, token)
+		endif
+	elseif(strlen(token) > 0)
+		JSON_AddJSON(jsonID, jsonPath, FormulaParser(token))
+	endif
+
+	return jsonID
+End
+
+// @brief add escape characters to a path element
+Function/S EscapeJsonPath(str)
+	String str
+
+	return ReplaceString("/", str, "~1")
+End
```

Comment:
Can it be that this function should actually be part of the JSON XOP?

Reply:
The function only escapes names of the elements. So I think, it is the user's decision where the path ends and where the element starts. Besides that, that is very primitive escaping and straight forward imho. So probably no.

Reply:
Okay fine. Remember you can pass `esc=0/1` to `JSON_GetKeys` to tune the escaping.

Reply:
[JSON_GetKeys](https://docs.byte-physics.de/json-xop/operations.html#get-keys) seems to be escaping by default. Should I turn that off using esc=1? - It is probably more readable turned off.

Reply:
sorry. It does not escape by default.

Reply:
I think it will not get prettier using the `esc` option:

```igorpro
	ASSERT(JSONtype == JSON_OBJECT, "Topmost element needs to be an object")
	WAVE/T operations = JSON_GetKeys(jsonID, jsonPath, esc = 1)
	ASSERT(DimSize(operations, ROWS) == 1, "Only one operation is allowed")
	jsonPathOld = jsonPath
	jsonPath += "/" + operations[0]
	WAVE/T operations = JSON_GetKeys(jsonID, jsonPathOld, esc = 0)
```

Reply:
and btw, syntax highlighting of such snippets is not working.